Fix copy to clipboard STA thread issue & Support retry for copy & Async build-in shortcut model & Fix build shortcuts text replace issue & Fix startup window hide issue

Jack251970 opened the pull request:
# Fix copy to clipboard STA thread issue

Fix `CopyToClipboard` function in JsonRPCAPI for JsonRPCV2 plugins: we should use STA thread to execute clipboard options.

Close #3071.

# Add retry for clipboard copy

Sometimes clipboard is locked, we need to wait and retry.

# Async buildin shortcut model

Since new clipboard function need async, we should add support for asynchronous build-in shortcut model.

# Fix query update issues

### Fix build shortcuts text replace issue

Fix build shortcuts text replace issue from jjw24 and remove global query delay

![image](https://github.com/user-attachments/assets/df6b0d62-6268-42b3-aab0-026ea4f58793)

### Fix possible ui thread issue

Check dispatcher access before invoking dispatch to fix possible ui thread issue

# Fix startup window hide issue

Fix issue that `HotkeyMapper` initialization will steal focus from main window which can cause main window forces to hide during startup.

# Test

* Copy to clipboard (text, file) on csharp plugins
* Copy to clipboard (text, file) on JsonRPCV2 plugins
* Get text from clipboard
* Get text from clipboard asynchronously
* Builtin shortcuts are replaced when we query them

Yusyuriv commented:
I am very against this. This is the core, most used, functionality, and it should be available to everyone. Plugins shouldn't need to reinvent the wheel every time.

Jack251970 commented:
> I am very against this. This is the core, most used, functionality, and it should be available to everyone. Plugins shouldn't need to reinvent the wheel every time. Especially considering Node and Python don't provide a way to do this. I remember seeing one plugin that packed another binary with itself just so it could copy text, which is absolutely ridiculous.

Well, could you please give me the sample to test?

And I think this issue only happens for JsonRPCv2 plugins. I am not sure.

Yusyuriv commented:
[demo-plugin.zip](https://github.com/user-attachments/files/19102747/demo-plugin.zip)

After installing this plugin, type `rpc2`. It will have one result. Select that result. You'll get a message box saying if the text was copied successfully or not.

Jack251970 commented:
> [demo-plugin.zip](https://github.com/user-attachments/files/19102747/demo-plugin.zip)
> 
> After installing this plugin, type `rpc2`. It will have one result. Select that result. You'll get a message box saying if the text was copied successfully or not.

Many thanks! Sorry, I don't see this as a bug unique to the V2 plugin. In that case, the bug could be fixed.

Jack251970 commented:
> [demo-plugin.zip](https://github.com/user-attachments/files/19102747/demo-plugin.zip)
> 
> After installing this plugin, type `rpc2`. It will have one result. Select that result. You'll get a message box saying if the text was copied successfully or not.

Sorry I cannot open your demo plugin. I have used 1.19.5 release version of FL (master branch) and downloaded embedded node.js environment from FL but I got this issue:

![image](https://github.com/user-attachments/assets/5cf683c9-9db0-427c-a875-770a7017df47)

Could you please fix it?

Yusyuriv commented:
Could you please install a [newer version of NodeJS (v22)](https://nodejs.org/en), specify it in settings (General — Node.js Path) and try again? Flow currently uses very old version that has reached end-of-life already.

Jack251970 commented:
> Could you please install a [newer version of NodeJS (v22)](https://nodejs.org/en), specify it in settings (General — Node.js Path) and try again? Flow currently uses very old version that has reached end-of-life already.

Thanks, I think I have fixed it. Now could you please help me check if copy to clipboard (directory) with JsonRPCV2 plugins can work?

Jack251970 commented:
> > Could you please install a [newer version of NodeJS (v22)](https://nodejs.org/en), specify it in settings (General — Node.js Path) and try again? Flow currently uses very old version that has reached end-of-life already.
> 
> Thanks, I think I have fixed it. Now could you please help me check if copy to clipboard (directory) with JsonRPCV2 plugins can work?

Nvm, I have checked it.

Jack251970 commented on Flow.Launcher.Infrastructure/UserSettings/Settings.cs:
```diff
@@ -253,7 +253,7 @@ public SearchPrecisionScore QuerySearchPrecision
         public ObservableCollection<BuiltinShortcutModel> BuiltinShortcuts { get; set; } = new()
         {
-            new BuiltinShortcutModel("{clipboard}", "shortcut_clipboard_description", Clipboard.GetText),
+            new BuiltinShortcutModel("{clipboard}", "shortcut_clipboard_description", () => Win32Helper.StartSTATaskAsync(Clipboard.GetText).Result),
```
I think either is ok?

Jack251970 replied:
> I think it's better to use GetAwaiter().GetResult().

This cannot work here, so I choose to keep as it was

taooceros replied:
what do you mean by not work here?

Jack251970 replied:
Changing `() => Win32Helper.StartSTATaskAsync(Clipboard.GetText).Result` to `Win32Helper.StartSTATaskAsync(Clipboard.GetText).GetAwaiter().GetResult` cannot work, the result is always empty

taooceros replied:
It makes no sense to me. They should perform the same...

https://stackoverflow.com/questions/17284517/is-task-result-the-same-as-getawaiter-getresult

taooceros replied:
I will test it.

taooceros commented on Flow.Launcher.Infrastructure/UserSettings/Settings.cs:
```diff
@@ -253,7 +253,7 @@ public SearchPrecisionScore QuerySearchPrecision
         public ObservableCollection<BuiltinShortcutModel> BuiltinShortcuts { get; set; } = new()
         {
-            new BuiltinShortcutModel("{clipboard}", "shortcut_clipboard_description", Clipboard.GetText),
+            new BuiltinShortcutModel("{clipboard}", "shortcut_clipboard_description", () => Win32Helper.StartSTATaskAsync(Clipboard.GetText).Result),
```
Actually how do you think about making the BuiltinShortcutModel supports Async operation?

Jack251970 replied:
Oh yeah, it is a good idea

Jack251970 replied:
Done

Jack251970 commented:
@taooceros I added asynchronous build shortcut model called `AsyncBuildinShortcutModel` and I changed related `ConstructQuery` to `ConstructQueryAsync`. Please review.

One thing I want to make sure is that I have removed `Application.Current.Dispatcher.Invoke(() =>` in original `ConstructQuery` function because I do not think there are any operations which need dispatcher, could you please help me double check this? Thanks!

jjw24 commented on Flow.Launcher.Infrastructure/Win32Helper.cs:
```diff
@@ -322,6 +324,78 @@ internal static HWND GetWindowHandle(Window window, bool ensure = false)
+        #region STA Thread
+
+        /*
+        Found on https://github.com/files-community/Files
```
```suggestion
        Inspired by https://github.com/files-community/Files code on STA Thread handling.
```

Jack251970 commented:
@taooceros Any time to take a look?

jjw24 commented:
Testing against current dev branch, there is a bit of delay in results update. As seen here in the gifs, when i type 'c' 'h' 'r', the pr version takes slightly longer time to update the results- looking at the first result 'Administrative Tools' takes a slight second to change to 'Google Chrome'.

This PR:
![pr-branch-chrome-delay](https://github.com/user-attachments/assets/c1fbf825-997c-4f28-bcab-78e1c298c673)

Dev:
![dev-branch-chrome-no-delay](https://github.com/user-attachments/assets/7f68a5ec-ac1d-4061-a9ef-cac41eaf6555)

Using 10.0.26100.3775

Log for PR:
[2025-04-27.txt](https://github.com/user-attachments/files/19927546/2025-04-27.txt)

Log for Dev
[2025-04-27.txt](https://github.com/user-attachments/files/19927557/2025-04-27.txt)

Jack251970 commented:
@jjw24  That should be normal behaviour. This PR will not start the next result update until last query is cancelled. See `await _updateLock.WaitAsync(CancellationToken.None);`

jjw24 commented:
This is causing wrong result trigger for fast typers. The behaviour of dev is better. What issue are we trying to resolve changing this behaviour?

Jack251970 commented:
> This is causing wrong result triggered for fast typers. The behaviour of dev is better. What issue are we trying to resolve changing this behaviour?

Now I enlarge the maximum number of update flows to six. And I think we can achieve a balance between performance and stability in that way.

Jack251970 commented:
> This is causing wrong result triggered for fast typers. The behaviour of dev is better. What issue are we trying to resolve changing this behaviour?

Originally, I think we should ensure the update flow cancellation and update one by one, but it seems it can lead to the performance problem as you described above.

jjw24 commented:
Can you please pull this one out of this PR and into its own so we can merge the rest of this for release. We can then look further into resolving and testing the results update change in a seperate PR.

Jack251970 commented:
> Can you please pull this one out of this PR and into its own so we can merge the rest of this for release. We can then look further into resolving and testing the results update change in the seperate PR.

Please check. I removed update lock from this PR.

Jack251970 commented:
@jjw24 Fixed result update issue in ebaffc6.

jjw24 commented:
The delay in result update is gone, but I am intermittently hitting no results returned from Programs plugin after heavy usage. When this happens, the log does not show the cost for querying Programs plugin.

Jack251970 commented:
> The delay in result update is gone, but I am intermittently hitting no results returned from Programs plugin after heavy usage. When this happens, the log does not show the cost for querying Programs plugin.

Have you installed the latest version of this PR? I added full logs here, and you can tracked to all operations of result updates.

jjw24 commented:
The log basically just doesn't have any indication of Programs plugin being queried so I didn't see the point of attaching the log. I will see if I can reliably reproduce the issue.

jjw24 commented:
The clear logs function in About page freezes the entire settings window. Could you check if reproducible on your end please.

Jack251970 commented:
> The clear logs function in About page freezes the entire settings window. Could you check if reproducible on your end please.

I cannot reproduce this on my side. Clear log works fine for me.

Jack251970 commented:
> The clear logs function in About page freezes the entire settings window. Could you check if reproducible on your end please.

Now I can reproduce it with `clear cache` option. Will fix it soon.

jjw24 commented:
Could you please still pull out the results update fix into its own PR because I am still not confident enough with the change (still getting weird occasional missing Programs plugin results I cannot reproduce reliably) and it needs to be tested and captured on its own separately.

jjw24 commented:
Can you reproduce the clear logs function freeze in this PR?

Jack251970 commented:
> Can you reproduce the clear logs function freeze in this PR?

No, the only change I made is setting recursive to false and I have reverted this in 
[Use recursive false](https://github.com/Flow-Launcher/Flow.Launcher/pull/3499/commits/c9e85a1888f0936342ea74121c7e070ff7322777)

jjw24 commented:
hm I have it happen on both laptops and both laptop's dev build had no issues that's why I thought it's related to this PR, but I can't reproduce it now. Let's not worry about it then if you haven't made a change in this area and I can't reliably reproduce.

Jack251970 commented:
> Could you please still pull out the results update fix into its own PR because I am still not confident enough with the change (still getting weird occasional missing Programs plugin results I cannot reproduce reliably) and it needs to be tested and captured on its own separately.

Before starting this, I would like to ask you two things:

- When the Program plugin cannot be queried, can other plugins be queried?
- In `QueryResultsAsync` function, I added Debug-level logs for the entire process from query input to plugin query, that is, when you start entering something, Flow will start outputting logs. Are you sure there is no log at all for the Program plugin that cannot be queried?

Jack251970 commented:
@jjw24 I reverted changes about the update logic and please check

jjw24 commented:
>When the Program plugin cannot be queried, can other plugins be queried?

Other plugins like Explorer can be queried.

>In QueryResultsAsync function, I added Debug-level logs for the entire process from query input to plugin query, that is, when you start entering something, Flow will start outputting logs. Are you sure there is no log at all for the Program plugin that cannot be queried?

Debug turned on, no logs for Programs plugin and logs for others like Explorer plugin.

taooceros commented on Flow.Launcher/ViewModel/MainViewModel.cs:
```diff
@@ -1331,36 +1371,56 @@ private Query ConstructQuery(string queryText, IEnumerable<CustomShortcutModel>
-                            e);
+                        else if (shortcut is AsyncBuiltinShortcutModel asyncShortcut)
+                        {
+                            expansion = App.JTF.Run(() => asyncShortcut.ExpandAsync());
```
let make it real async (or else it does not really makes sense to create an async shortcut)

jjw24 reviewed (approved):
Tested prior to latest commit:
- result update has no issues
- querying response is fast
- no missing Programs plugin results after intense use
- copying file, paths, bookmark urls

Jack251970 commented:
> Tested prior to latest commit:
> 
> * result update has no issues
> * querying response is fast
> * no missing Programs plugin results after intense use
> * copying file, paths, bookmark urls

Good!

jjw24 commented:
Where is the PR for the changes you pulled out?

Jack251970 commented:
> Where is the PR for the changes you pulled out?

I will handle it now. I have not created PR yet.

Jack251970 commented:
> Where is the PR for the changes you pulled out?

Please check #3502.
